lib/PreviewProvider.js

```diff
@@ -0,0 +1,239 @@
+      if (val.length !== 0 && key.length !== 0) {
+        safeQueryParam = safeQueryParam + key + "=" + val + "&";
+      }
+    });
+    url.search = safeQueryParam.slice(0, safeQueryParam.length - 1);
+    url.hash = "";
+
+    // remove extra slashes in pathname and construct back a safe pathname
+    if (url.pathname) {
+      let pathItems = url.pathname.split("/");
+      let safePathItems = pathItems.filter(item => item.length !== 0);
+      let safePath = "/";
+      url.pathname = safePath + safePathItems.join("/");
+    }
+
+    // if the url doesn't have a protocol, use https
+    if (!url.protocol) {
+      url.protocol = "https:";
+    }
+
+    return url.toString();
+  },
+
+  /**
+    * Create a key based on a URL in order to dedupe sites
+    */
+  _createKey(url) {
+    url = new URL(url);
+    let key = url.host.replace(/www\.?/, "");
+    key = key + url.pathname;
+    return key;
+  },
+
+  /**
+    * Canonicalize urls by deduping them, then sanitizing them
+    */
+  _dedupeURLs(sites) {
+    let dedupedSites = new Map();
+    sites.forEach(site => {
+      let urlkey = this._createKey(site.url);
+      if (!dedupedSites.has(urlkey)) {
+        dedupedSites.set(urlkey, site);
+      } else {
+        // if two sites come in with 'http' and 'https', pick 'https'
+        if (!dedupedSites.get(urlkey).url.contains("https")) {
+          dedupedSites.delete(urlkey);
+          dedupedSites.set(urlkey, site);
+        }
+      }
+    });
+    let freshSites = [];
+    dedupedSites.forEach((site) => {
+      // sanitize and normalize each site
+      site.url = this._sanitizeURL(site.url);
+      freshSites.push(site);
+    });
+    return freshSites;
+  },
+
+  /**
+    * Filter cached links out, and store their embedly data
+    */
+  getCachedLinks(links) {
+    let cachedLinks = [];
+    let sites = this._dedupeURLs(links);
+    sites.forEach(link => {
+      if (link && ss.storage.embedlyData[link.url] && link.url === ss.storage.embedlyData[link.url].url) {
+        cachedLinks.push(ss.storage.embedlyData[link.url]);
+      }
+    });
+    return cachedLinks;
+  },
+
+  /**
+   * Filter out new links and request them from embedly
+   */
+  saveNewLinks: Task.async(function*(links) {
+    let newLinks = links.filter(link => link && !ss.storage.embedlyData[link.url]);
+    let requestQueue = [];
+    while (newLinks.length !== 0) {
+      // we have some new links we need to fetch the embedly data for, put them on the queue
+      requestQueue.push(newLinks.splice(0, EMBEDLY_PROXY_MAX_LINKS));
+    }
+    // for each bundle of 25 links, create a new request to embedly
+    requestQueue.forEach(requestBundle => {
+      this._fetchAndCache(requestBundle);
+    });
+  }),
+
+  /**
+   * Makes the necessary requests to embedly to get data and store it in the cache
+   */
+  _fetchAndCache: Task.async(function*(newLinks) {
+    let json = null;
+    let sites = this._dedupeURLs(newLinks.filter(this._URLFilter(URL_FILTERS)));
```

Comment:
we might still be sending multiple duplicates to our embedly proxy if the urls happen to not be in the expected batch